src/renderers/shared/fiber/__tests__/ReactIncrementalTriangle-test.js

```diff
@@ -0,0 +1,325 @@
+    };
+  }
+
+  const TOGGLE = 'TOGGLE';
+  function toggle(childIndex) {
+    return {
+      type: TOGGLE,
+      childIndex,
+    };
+  }
+
+  function TriangleSimulator() {
+    let triangles = [];
+    let leafTriangles = [];
+    let yieldAfterEachRender = false;
+    class Triangle extends React.Component {
+      constructor(props) {
+        super();
+        this.index = triangles.length;
+        triangles.push(this);
+        if (props.depth === 0) {
+          this.leafIndex = leafTriangles.length;
+          leafTriangles.push(this);
+        }
+        this.state = {isActive: false};
+      }
+      activate() {
+        if (this.props.depth !== 0) {
+          throw new Error('Cannot activate non-leaf component');
+        }
+        ReactNoop.performAnimationWork(() => {
+          this.setState({isActive: true});
+        });
+      }
+      deactivate() {
+        if (this.props.depth !== 0) {
+          throw new Error('Cannot deactivate non-leaf component');
+        }
+        ReactNoop.performAnimationWork(() => {
+          this.setState({isActive: false});
+        });
+      }
+      shouldComponentUpdate(nextProps, nextState) {
+        return (
+          this.props.counter !== nextProps.counter ||
+          this.state.isActive !== nextState.isActive
+        );
+      }
+      render() {
+        if (yieldAfterEachRender) {
+          ReactNoop.yield(this);
+        }
+        const {counter, depth} = this.props;
+        if (depth === 0) {
+          if (this.state.isActive) {
+            return <span prop={'*' + counter + '*'} />;
+          }
+          return <span prop={counter} />;
+        }
+        return [
+          <Triangle key={1} counter={counter} depth={depth - 1} />,
+          <Triangle key={2} counter={counter} depth={depth - 1} />,
+          <Triangle key={3} counter={counter} depth={depth - 1} />,
+        ];
+      }
+    }
+
+    let appInstance;
+    class App extends React.Component {
+      state = {counter: 0};
+      interrupt() {
+        // Triggers a restart from the top.
+        ReactNoop.performAnimationWork(() => {
+          this.forceUpdate();
+        });
+      }
+      setCounter(counter) {
+        const currentCounter = this.state.counter;
+        this.setState({counter});
+        return currentCounter;
+      }
+      render() {
+        appInstance = this;
+        return <Triangle counter={this.state.counter} depth={3} />;
+      }
+    }
+
+    const depth = 3;
+
+    let keyCounter = 0;
+    function reset(nextStep = 0) {
+      triangles = [];
+      leafTriangles = [];
+      // Remounts the whole tree by changing the key
+      ReactNoop.render(<App depth={depth} key={keyCounter++} />);
+      ReactNoop.flush();
+      treeIsConsistent();
+      return appInstance;
+    }
+
+    reset();
+    const totalChildren = leafTriangles.length;
+    const totalTriangles = triangles.length;
+
+    function treeIsConsistent(activeTriangle, counter) {
```

Comment:
Minor nit: maybe `assertConsistentTree` or something that makes it clearer this is throwing.